src/stripeLogsView.ts

```diff
@@ -1,15 +1,164 @@
+      this.setViewState(ViewState.Idle);
+    }
+  }
+
+  stopLogsStreaming() {
+    this.stripeClient.endCLIProcess(CLICommand.LogsTail);
+    this.cleanupStreams();
+    this.setViewState(ViewState.Idle);
+  }
+
   buildTree(): Promise<StripeTreeItem[]> {
-    const logStreamItem = new StripeTreeItem('Start API logs streaming', 'openLogsStreaming');
-    logStreamItem.setIcon({
-      dark: Resource.ICONS.dark.terminal,
-      light: Resource.ICONS.light.terminal,
-    });
+    const streamingControlItemArgs = (() => {
+      switch (this.viewState) {
+        case ViewState.Idle:
+          return {
+            label: 'Start streaming API logs',
+            command: 'startLogsStreaming',
+            iconId: 'play-circle',
+          };
+        case ViewState.Loading:
+          return {
+            label: 'Starting streaming API logs...',
+            command: 'stopLogsStreaming',
+            iconId: 'loading',
+          };
+        case ViewState.Streaming:
+          return {
+            label: 'Stop streaming API logs',
+            command: 'stopLogsStreaming',
+            iconId: 'stop-circle',
+          };
+      }
+    })();
+
+    const streamingControlItem = this.createItemWithCommand(streamingControlItemArgs);
+
+    const treeItems = [streamingControlItem];
+
+    if (this.logTreeItems.length > 0) {
+      const logsStreamRootItem = new StripeTreeItem('Recent logs');
+      logsStreamRootItem.children = this.logTreeItems;
+      logsStreamRootItem.expand();
+      treeItems.push(logsStreamRootItem);
+    }
+
+    return Promise.resolve(treeItems);
+  }
+
+  private createItemWithCommand({
+    label,
+    command,
+    iconId,
+  }: {
+    label: string;
+    command?: string;
+    iconId?: string;
+  }) {
+    const item = new StripeTreeItem(label, command);
+    if (iconId) {
+      item.iconPath = new ThemeIcon(iconId);
+    }
+    return item;
+  }
+
+  private async setupStreams() {
+    const stripeLogsTailProcess = await this.stripeClient.getOrCreateCLIProcess(
+      CLICommand.LogsTail,
+      ['--format', 'JSON'],
+    );
+    if (!stripeLogsTailProcess) {
+      throw new Error('Failed to start `stripe logs tail` process');
+    }
+
+    // The CLI lets you know that streaming is ready via stderr
+    if (!this.logsStderrStream) {
+      await new Promise<void>((resolve) => {
+        this.logsStderrStream = new stream.Writable({
+          write: (chunk, _, callback) => {
+            const label = chunk.toString();
+            if (label.includes('Ready!')) {
+              resolve();
+            }
+            callback();
+          },
+        });
+        stripeLogsTailProcess.stderr.pipe(this.logsStderrStream);
+      });
+    }
+
+    if (!this.logsStdoutStream) {
+      this.logsStdoutStream = new stream.Writable({
+        write: (chunk, _, callback) => {
+          try {
+            const logObject = JSON.parse(chunk.toString());
+            if (logObject && typeof logObject === 'object') {
```

Comment:
probably want to be a bit more defensive in looking for objects that smell like what we need.